Fix: cron completion race condition

rhell4 opened the pull request:
This fixes a race condition where the completion select takes too long to return the results and in that time the user has triggered an event observer that has already inserted the completion.

rhell4 committed:
Fix: cron completion race condition

This fixes a race condition where the completion select takes too long
to return the results and in that time the user has triggered an event
observer that has already inserted the completion.

matthewhilton commented on classes/local/allocation.php:
```diff
@@ -513,7 +511,14 @@ public static function fix_user_enrolments(?int $programid, ?int $userid): void
+        $rs = $DB->get_recordset_sql($sql, $params);
+        foreach ($rs as $completion) {
+            $params = ['itemid' => $completion->itemid, 'allocationid' => $completion->allocationid];
+            if (!$DB->record_exists('enrol_programs_completions', $params)) {
```
Might be worth adding a comment here explaining why this is done.

matthewhilton commented on classes/local/allocation.php:
```diff
@@ -545,7 +548,14 @@ public static function fix_user_enrolments(?int $programid, ?int $userid): void
+        $rs = $DB->get_recordset_sql($sql, $params);
+        foreach ($rs as $completion) {
+            $params = ['itemid' => $completion->itemid, 'allocationid' => $completion->allocationid];
+            if (!$DB->record_exists('enrol_programs_completions', $params)) {
```
Same here might be worth adding a comment explaining why

matthewhilton reviewed (commented):
Looks good!
Looking at the code the issue should be fixed, but wonder if there is a way to test this (either manual or a unit test) ?

Also, since there isn't CI in the repo, could you confirm that the unit tests still pass (just want to be sure we aren't breaking the logic accidentally)

rhell4 commented:
Hey @matthewhilton I've added some comments and changed some variable names as they were causing param errors in the loops. Can confirm unit tests are passing now.

It's a bit hard to test this scenario in the unit test since it requires two processes to be happening at the same time. Even manually is a bit hard since it requires a user completing a course at the exact same moment one of those queries is being called and requires a pretty big DB to slow down that query.

matthewhilton commented:
Thanks @rhell4 - Agreed re. testing, its a bit of a tricky edge case - not an issue. We'll know this works out in the wild if we see less/no scheduled task fails.

Good to see unit tests passing - so at least we know we haven't broken anything further :smile:
